src/main/java/net/javadiscord/javabot2/systems/moderation/MessageCache.java

```diff
@@ -0,0 +1,82 @@
+package net.javadiscord.javabot2.systems.moderation;
+
+import net.javadiscord.javabot2.Bot;
+import org.javacord.api.entity.message.Message;
+import org.javacord.api.entity.message.MessageAuthor;
+import org.javacord.api.event.message.MessageCreateEvent;
+import org.javacord.api.listener.message.MessageCreateListener;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.LinkedList;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Caches incoming messages and cleans it out when needed.
+ */
+public class MessageCache implements MessageCreateListener {
+
+	private final ConcurrentHashMap<MessageAuthor, Instant> lastModifications;
+	private final ConcurrentHashMap<MessageAuthor, LinkedList<Message>> cache;
+
+    /**
+     * Creates the cache.
+     */
+	public MessageCache() {
+		lastModifications = new ConcurrentHashMap<>();
+		cache = new ConcurrentHashMap<>();
```

Comment:
You're constructing these so that the outer Map is thread-safe, but in `add`, you're modifying the contents of a linked list within the map. That is not threadsafe and will lead to ConcurrentModificationException.